File: androidApp/src/main/kotlin/dev/lildua/oddly/ui/components/Surfaces.kt
package dev.lildua.oddly.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.lildua.oddly.ui.theme.OddlyGradients
import dev.lildua.oddly.ui.theme.OddlyTheme

/** The standard content container: rounded, slightly lifted, hairline border. */
@Composable
fun OddlyCard(
    modifier: Modifier = Modifier,
    shape: Shape = RoundedCornerShape(20.dp),
    background: Color = MaterialTheme.colorScheme.surface,
    borderColor: Color = MaterialTheme.colorScheme.outline.copy(alpha = 0.6f),
    contentPadding: Dp = 18.dp,
    onClick: (() -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val scale by animateFloatAsState(if (pressed && onClick != null) 0.985f else 1f, label = "card-scale")

    Column(
        modifier = modifier
            .fillMaxWidth()
            .scale(scale)
            .clip(shape)
            .background(background)
            .border(BorderStroke(1.dp, borderColor), shape)
            .then(
                if (onClick != null) {
                    Modifier.clickable(
                        interactionSource = interaction,
                        indication = null,
                        onClick = onClick,
                    )
                } else {
                    Modifier
                },
            )
            .padding(contentPadding),
        content = content,
    )
}

/** Card variant filled with a brush — used for hero and category surfaces. */
@Composable
fun GradientCard(
    brush: Brush,
    modifier: Modifier = Modifier,
    shape: Shape = RoundedCornerShape(20.dp),
    borderColor: Color = Color.Transparent,
    contentPadding: Dp = 18.dp,
    onClick: (() -> Unit)? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .background(brush)
            .border(BorderStroke(1.dp, borderColor), shape)
            .then(
                if (onClick != null) {
                    Modifier.clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        onClick = onClick,
                    )
                } else {
                    Modifier
                },
            )
            .padding(contentPadding),
        content = content,
    )
}

/** Text painted with a gradient — the "1%" wordmark and hero numbers. */
@Composable
fun GradientText(
    text: String,
    style: TextStyle,
    modifier: Modifier = Modifier,
    brush: Brush = OddlyGradients.brandText,
) {
    Text(text = text, style = style.copy(brush = brush), modifier = modifier)
}

/** Small uppercase label that heads a section, e.g. "GỢI Ý", "PHẦN THƯỞNG". */
@Composable
fun SectionLabel(
    text: String,
    modifier: Modifier = Modifier,
    color: Color = OddlyTheme.palette.textTertiary,
) {
    Text(
        text = text.uppercase(),
        style = MaterialTheme.typography.labelSmall,
        color = color,
        modifier = modifier,
    )
}

/** A rounded progress track filled with a gradient. */
@Composable
fun GradientProgressBar(
    progress: Float,
    modifier: Modifier = Modifier,
    height: Dp = 8.dp,
    brush: Brush = OddlyGradients.progress,
    trackColor: Color = OddlyTheme.palette.surfaceHighest,
) {
    val animated by animateFloatAsState(progress.coerceIn(0f, 1f), label = "progress")
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(height)
            .clip(RoundedCornerShape(percent = 50))
            .background(trackColor),
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth(animated)
                .height(height)
                .clip(RoundedCornerShape(percent = 50))
                .background(brush),
        )
    }
}

/** Screen header with an optional back affordance and trailing slot. */
@Composable
fun OddlyTopBar(
    title: String,
    modifier: Modifier = Modifier,
    onBack: (() -> Unit)? = null,
    trailing: (@Composable () -> Unit)? = null,
) {
    val palette = OddlyTheme.palette
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 20.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        if (onBack != null) {
            Box(
                modifier = Modifier
                    .size(38.dp)
                    .clip(RoundedCornerShape(percent = 50))
                    .background(palette.surfaceElevated)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        onClick = onBack,
                    ),
                contentAlignment = Alignment.Center,
            ) {
                OddlyIcon(OddlyIcon.ChevronLeft, size = 20.dp, tint = palette.textPrimary)
            }
        }
        Text(
            text = title,
            style = MaterialTheme.typography.headlineSmall.copy(fontWeight = FontWeight.Bold),
            color = palette.textPrimary,
            modifier = Modifier.weight(1f),
        )
        trailing?.invoke()
    }
}

/** A pill-shaped tag: difficulty, category, estimated time. */
@Composable
fun OddlyChip(
    text: String,
    modifier: Modifier = Modifier,
    accent: Color = OddlyTheme.palette.textSecondary,
    leadingEmoji: String? = null,
    selected: Boolean = false,
    onClick: (() -> Unit)? = null,
) {
    val palette = OddlyTheme.palette
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(percent = 50))
            .background(if (selected) accent.copy(alpha = 0.2f) else palette.surfaceElevated)
            .border(
                BorderStroke(1.dp, if (selected) accent.copy(alpha = 0.7f) else Color.Transparent),
                RoundedCornerShape(percent = 50),
            )
            .then(
                if (onClick != null) {
                    Modifier.clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        onClick = onClick,
                    )
                } else {
                    Modifier
                },
            )
            .padding(horizontal = 12.dp, vertical = 7.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        leadingEmoji?.let {
            Text(it, style = MaterialTheme.typography.labelMedium)
        }
        Text(
            text = text,
            style = MaterialTheme.typography.labelMedium,
            color = if (selected) accent else palette.textSecondary,
        )
    }
}

File: iosApp/iosApp/Components/Surfaces.swift
import SwiftUI

/// The standard content container: rounded, slightly lifted, hairline border.
struct OddlyCard<Content: View>: View {
    @Environment(\.palette) private var palette

    var cornerRadius: CGFloat = OddlyRadius.large
    var background: Color?
    var borderColor: Color?
    var contentPadding: CGFloat = 18
    var action: (() -> Void)?
    @ViewBuilder var content: () -> Content

    var body: some View {
        let shape = RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
        let surface = VStack(alignment: .leading, spacing: 0) { content() }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(contentPadding)
            .background(background ?? palette.surface, in: shape)
            .overlay(shape.stroke(borderColor ?? palette.outline.opacity(0.6), lineWidth: 1))

        if let action {
            Button(action: action) { surface }
                .buttonStyle(PressableStyle(pressedScale: 0.985))
        } else {
            surface
        }
    }
}

/// Card variant filled with a gradient — used for hero and category surfaces.
struct GradientCard<Content: View>: View {
    var gradient: LinearGradient
    var cornerRadius: CGFloat = OddlyRadius.large
    var borderColor: Color = .clear
    var contentPadding: CGFloat = 18
    var action: (() -> Void)?
    @ViewBuilder var content: () -> Content

    var body: some View {
        let shape = RoundedRectangle(cornerRadius: cornerRadius, style: .continuous)
        let surface = VStack(alignment: .leading, spacing: 0) { content() }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(contentPadding)
            .background(gradient, in: shape)
            .overlay(shape.stroke(borderColor, lineWidth: 1))

        if let action {
            Button(action: action) { surface }
                .buttonStyle(PressableStyle(pressedScale: 0.985))
        } else {
            surface
        }
    }
}

/// Text painted with a gradient — the "1%" wordmark and hero numbers.
struct GradientText: View {
    let text: String
    var font: Font
    var gradient: LinearGradient = OddlyGradients.brandText
    var tracking: CGFloat = 0

    init(_ text: String, font: Font, gradient: LinearGradient = OddlyGradients.brandText, tracking: CGFloat = 0) {
        self.text = text
        self.font = font
        self.gradient = gradient
        self.tracking = tracking
    }

    var body: some View {
        Text(text)
            .font(font)
            .tracking(tracking)
            .multilineTextAlignment(.center)
            .foregroundStyle(gradient)
    }
}

/// Small uppercase label that heads a section, e.g. "GỢI Ý", "PHẦN THƯỞNG".
struct SectionLabel: View {
    @Environment(\.palette) private var palette

    let text: String
    var color: Color?

    init(_ text: String, color: Color? = nil) {
        self.text = text
        self.color = color
    }

    var body: some View {
        Text(text.uppercased())
            .font(OddlyFont.labelSmall)
            .tracking(0.5)
            .foregroundStyle(color ?? palette.textTertiary)
    }
}

/// A rounded progress track filled with a gradient.
struct GradientProgressBar: View {
    @Environment(\.palette) private var palette

    let progress: Double
    var height: CGFloat = 8
    var gradient: LinearGradient = OddlyGradients.progress
    var trackColor: Color?

    var body: some View {
        GeometryReader { geometry in
            let clamped = min(max(progress, 0), 1)
            ZStack(alignment: .leading) {
                Capsule().fill(trackColor ?? palette.surfaceHighest)
                Capsule()
                    .fill(gradient)
                    .frame(width: geometry.size.width * clamped)
            }
        }
        .frame(height: height)
        .animation(.easeOut(duration: 0.35), value: progress)
    }
}

/// Screen header with an optional back affordance and trailing slot.
struct OddlyTopBar<Trailing: View>: View {
    @Environment(\.palette) private var palette

    let title: String
    var onBack: (() -> Void)?
    @ViewBuilder var trailing: () -> Trailing

    var body: some View {
        HStack(spacing: 12) {
            if let onBack {
                CircleIconButton(.chevronLeft, diameter: 38, iconSize: 20, action: onBack)
            }
            Text(title)
                .font(OddlyFont.headlineSmall.weight(.bold))
                .foregroundStyle(palette.textPrimary)
                .frame(maxWidth: .infinity, alignment: .leading)
            trailing()
        }
        .padding(.horizontal, 20)
        .padding(.vertical, 12)
    }
}

extension OddlyTopBar where Trailing == EmptyView {
    init(title: String, onBack: (() -> Void)? = nil) {
        self.init(title: title, onBack: onBack) { EmptyView() }
    }
}

/// A pill-shaped tag: difficulty, category, estimated time.
struct OddlyChip: View {
    @Environment(\.palette) private var palette

    let text: String
    var accent: Color?
    var leadingEmoji: String?
    var selected: Bool = false
    var action: (() -> Void)?

    var body: some View {
        let tint = accent ?? palette.textSecondary
        let pill = HStack(spacing: 6) {
            if let leadingEmoji {
                Text(leadingEmoji).font(OddlyFont.labelMedium)
            }
            Text(text)
                .font(OddlyFont.labelMedium)
                .foregroundStyle(selected ? tint : palette.textSecondary)
                .lineLimit(1)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 7)
        .background(selected ? tint.opacity(0.2) : palette.surfaceElevated, in: Capsule())
        .overlay(Capsule().stroke(selected ? tint.opacity(0.7) : .clear, lineWidth: 1))

        if let action {
            Button(action: action) { pill }
                .buttonStyle(PressableStyle())
        } else {
            pill
        }
    }
}

/// Section heading with an optional trailing action, e.g. "Xem tất cả".
struct SectionHeader: View {
    @Environment(\.palette) private var palette

    let title: String
    var actionText: String?
    var action: (() -> Void)?

    var body: some View {
        HStack {
            Text(title)
                .font(OddlyFont.titleMedium)
                .foregroundStyle(palette.textPrimary)
            Spacer()
            if let actionText, let action {
                TextAction(actionText, color: OddlyColors.purple, action: action)
            }
        }
    }
}

/// A switch tinted to the app's palette rather than the system accent.
struct OddlyToggle: View {
    @Binding var isOn: Bool

    var body: some View {
        Toggle("", isOn: $isOn)
            .labelsHidden()
            .tint(OddlyColors.success)
    }
}
